tools/fuzzer.cpp

```diff
@@ -0,0 +1,403 @@
+////////////////////////////////////////////////////////////////////////////////
+/// @brief Library to build up VPack documents.
+///
+/// DISCLAIMER
+///
+/// Copyright 2022 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+/// @author Copyright 2022, ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+
+#include <charconv>
+#include <cmath>
+#include <iostream>
+#include <mutex>
+#include <random>
+#include <string>
+#include <thread>
+
+#include "velocypack/vpack.h"
+#include "velocypack/velocypack-exception-macros.h"
+
+using namespace arangodb::velocypack;
+
+struct VPackFormat {};
+
+struct JSONFormat {};
+
+enum RandomBuilderAdditions {
+  ADD_ARRAY = 0,
+  ADD_OBJECT,
+  ADD_BOOLEAN,
+  ADD_STRING,
+  ADD_NULL,
+  ADD_UINT64,
+  ADD_INT64,
+  ADD_DOUBLE, // here starts values that are only for vpack
+  ADD_UTC_DATE,
+  ADD_BINARY,
+  ADD_EXTERNAL,
+  ADD_ILLEGAL,
+  ADD_MIN_KEY,
+  ADD_MAX_KEY,
+  ADD_MAX_VPACK_VALUE
+};
+
+struct KnownLimitValues {
+  static constexpr uint32_t maxDepth = 10;
+  static constexpr uint32_t utf81ByteFirstLowerBound = 0x00;
+  static constexpr uint32_t utf81ByteFirstUpperBound = 0x7F;
+  static constexpr uint32_t utf82BytesFirstLowerBound = 0xC0;
+  static constexpr uint32_t utf82BytesFirstUpperBound = 0xDF;
+  static constexpr uint32_t utf83BytesFirstLowerBound = 0xE0;
+  static constexpr uint32_t utf83BytesFirstUpperBound = 0xEF;
+  static constexpr uint32_t utf84BytesFirstLowerBound = 0xF0;
+  static constexpr uint32_t utf84BytesFirstUpperBound = 0xF7;
+  static constexpr uint32_t utf8CommonLowerBound = 0x80;
+  static constexpr uint32_t utf8CommonUpperBound = 0xBF;
+  static constexpr uint32_t minUtf8RandStringLength = 1;
+  static constexpr uint32_t maxUtf8RandStringLength = 1000;
+};
+
+struct RandomGenerator {
+  RandomGenerator(uint64_t seed) : mt64(seed) {}
+
+  std::mt19937_64 mt64; // this is 64 bits for generating uint64_t in ADD_UINT64 for velocypack
+};
+
+Slice nullSlice(Slice::nullSlice());
+
+static std::mutex mtx;
+
+static void usage(char* argv[]) {
+  std::cout << "Usage: " << argv[0] << " [FORMAT] [ITERATIONS] [THREADS] SEED"
```

Comment:
Usage text isn't up to date anymore. Took me a while to figure out how to actually invoke the fuzzer correctly.